// redux-modules/lib/moduleProvider/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createRegisterModule = require('./createRegisterModule');

var _createRegisterModule2 = _interopRequireDefault(_createRegisterModule);

var _storeShape = require('../utils/storeShape');

var _storeShape2 = _interopRequireDefault(_storeShape);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ModuleProvider = function (_Component) {
  _inherits(ModuleProvider, _Component);

  function ModuleProvider(props, context) {
    _classCallCheck(this, ModuleProvider);

    var _this = _possibleConstructorReturn(this, (ModuleProvider.__proto__ || Object.getPrototypeOf(ModuleProvider)).call(this, props, context));

    _this.store = props.store;
    _this.registerModule = (0, _createRegisterModule2.default)(props.store, props.combineReducers);
    if (typeof props.staticReducers !== 'undefined') {
      _this.registerModule(props.staticReducers);
    }
    return _this;
  }

  _createClass(ModuleProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        registerModule: this.registerModule,
        store: this.store
      };
    }
  }, {
    key: 'render',
    value: function render() {
      return _react.Children.only(this.props.children);
    }
  }]);

  return ModuleProvider;
}(_react.Component);

ModuleProvider.propTypes = {
  children: _propTypes2.default.element.isRequired,
  combineReducers: _propTypes2.default.func,
  // eslint-disable-next-line react/forbid-prop-types
  staticReducers: _propTypes2.default.object,
  store: _storeShape2.default.isRequired
};
ModuleProvider.childContextTypes = {
  registerModule: _propTypes2.default.func.isRequired,
  store: _storeShape2.default.isRequired
};
exports.default = ModuleProvider;
